components/service/glean/src/main/java/mozilla/components/service/glean/scheduler/MetricsPingScheduler.kt

```diff
@@ -6,77 +6,255 @@ package mozilla.components.service.glean.scheduler
+        // is more suitable for this task, because of the inherent drifting. See
+        // https://developer.android.com/reference/androidx/work/PeriodicWorkRequest.html
+        // for more details.
+        val workRequest = OneTimeWorkRequestBuilder<MetricsPingWorker>()
+            .addTag(MetricsPingScheduler.METRICS_PING_WORKER_TAG)
+            .setInitialDelay(millisUntilNextDueTime, AndroidTimeUnit.MILLISECONDS)
+            .build()
+
+        // Enqueue the work request: replace older requests if needed. This is to cover
+        // the odd case in which:
+        // - glean is killed, but the work request is still there;
+        // - glean restarts;
+        // - the ping is overdue and is immediately collected at startup;
+        // - a new work is scheduled for the next calendar day.
+        WorkManager.getInstance().enqueueUniqueWork(
+            MetricsPingScheduler.METRICS_PING_WORKER_TAG,
+            ExistingWorkPolicy.REPLACE,
+            workRequest)
+    }
+
+    /**
+     * Computes the amount of milliseconds until the next metrics ping due time.
      *
-     * @return true if the time interval has elapsed or if this is the first attempt to send a
-     *         metrics ping, otherwise false.
+     * @param sendTheNextCalendarDay whether or not to return the delay for today or tomorrow's
+     *        [dueHourOfTheDay]
+     * @param now the current datetime, a [Calendar] instance
+     * @param dueHourOfTheDay the due hour of the day, in the [0, 23] range.
+     * @return the milliseconds until the due hour: if current time is before the due
+     *         hour, then |dueHour - currentHour| is returned. If it's exactly on that hour,
+     *         then 0 is returned. Same if we're past the due hour.
      */
-    fun canSendPing(): Boolean {
-        val lastPingTime = try {
-            sharedPreferences.getLong(LAST_METRICS_PING_TIMESTAMP_KEY, 0L)
-        } catch (e: ClassCastException) {
-            // Intentionally doing nothing here, if LAST_METRICS_PING_TIMESTAMP_KEY's value is
-            // not a Long, then we will let lastPingTime be 0L for the purposes of comparison in
-            // order to allow the ping to be sent immediately
-            logger.error("MetricsPingScheduler last ping time stored was not a Long value")
-            0L
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    internal fun getMillisecondsUntilDueTime(
+        sendTheNextCalendarDay: Boolean,
+        now: Calendar = getCalendarInstance(),
+        dueHourOfTheDay: Int = DUE_HOUR_OF_THE_DAY
+    ): Long {
+        val nowInMillis = now.timeInMillis
+        val dueTime = getDueTimeForToday(now, dueHourOfTheDay)
+        val delay = dueTime.timeInMillis - nowInMillis
+        return when {
+            sendTheNextCalendarDay -> {
+                // We're past the `dueHourOfTheDay` in the current calendar day.
+                dueTime.add(Calendar.DAY_OF_MONTH, 1)
```

Comment:
This is a weird API, but it looks correct :)

Reply:
Heh, yeah, I agree!